### ISSUE TITLE
Line breaks sometimes happen immediately after an inline KaTeX block

### ISSUE THREAD

**Issue by akalin** [Khan/KaTeX#1233]
This is undesirable if e.g. a comma follow an inline equation.

This seems to happen to me most often on Chrome for Linux (Ubuntu 17.10), but I'm not 100% sure I haven't seen it on Chrome for Mac.

Example [gist](https://gist.github.com/akalin/15067e7ec3f7fbdd07d4fbb11b192918) with [display](https://rawgit.com/akalin/15067e7ec3f7fbdd07d4fbb11b192918/raw/330f631b36b60251ff2586908c8204a99732e9c6/test-0.9.0.html). If you adjust the right border of the browser window, you can get it to break between "A = (ε_0 + H)/2" and its following comma.

I'm not sure if KaTeX has much control over this behavior, but I figure I'd report it in case anyone has any ideas.

Screenshot:

![screenshot from 2018-03-20 23-36-23](https://user-images.githubusercontent.com/425576/37697132-495ca600-2c98-11e8-97ff-a1a2220f3cc4.png)

**Comment by akalin** [comment]
Writing the equation as "A = (ε_0 + H)/2\text{,}" does work in this case, but it would be nice not to have to do that. (Also, I'm not 100% sure that it sometimes doesn't break between the two anyway.)

**Comment by edemaine** [comment]
I've encountered this too, and tweaked [my own KaTeX preprocessor](https://github.com/edemaine/coauthor/blob/master/lib/formats.coffee#L430) to put following punctuation inside any math blocks.  This definitely works, as KaTeX doesn't support any line breaks within a math block. 

It'd be nice if there'd be a pure CSS workaround...  Alternatively, auto-render and a solution to #604 (which I'm imagining roughly following the code linked above) could implement this hacky solution.

**Comment by akalin** [comment]
Ah, cool, glad to hear this is a known thing. I was considering modifying auto-render, too, but I don't quite understand what #604 has to do with it. My idea would be as follows:

1) After autorender splits up a piece of text into a list of text and math blocks, but before it renders the math...
2) For each text block following an inline math block, if some option is set with a string of punctuation, or a regexp or something...
3) If the beginning of the text block matches the punctuation string/regexp, wrap it in a `\text{...}` fragment and append it to the previous inline math block. Looks like you use a nobr span; that might work, too.

Would this be something that would be useful to have in auto-render, or should I expect to just have it for myself?

**Comment by kevinbarabash** [comment]
I'm surprised/disappointed that browsers don't keep the punctuation on the same line unless it follows a space.

**Comment by Reedbeta** [comment]
I'm seeing this issue too. From a bit of experimentation in Firefox dev tools, it appears this is related to the use of `display: inline-block` on KaTeX spans. If I remove that property from `.katex .base`, then the browser no longer puts line breaks between inline math and punctuation.

However, removing `display: inline-block` also messes with the vertical spacing around math blocks (both display and inline), so that's not a complete fix in itself.

**Comment by vincentbernat** [comment]
Adding `.katex-html { white-space: nowrap; }` solves the issue, but it prevents inline formula to have a break inside the formula.

**Comment by ekzhang** [comment]
I was able to fix this issue by combining @vincentbernat's solution with a script that inserts zero-width spaces between each child of the `span.katex-html` element, which solves the problem of inline formulae not allowing breaks within the formula.

```javascript
try {
  katex.render(value, element, {
    displayMode,
    macros,
    throwOnError: false,
  });

  /**
   * Set `white-space` on the parent element to `nowrap`, and then insert
   * zero-width spaces between children to enable line breaks.
   *
   * Fix for: https://github.com/KaTeX/KaTeX/issues/1233
   */
  for (const span of element.getElementsByClassName("katex-html")) {
    span.setAttribute("style", "white-space: nowrap");
    const children = Array.from(span.children).slice(1);
    for (const child of children) {
      const spacer = document.createElement("span");
      spacer.innerText = "\u200b"; // Zero-width space
      spacer.style.whiteSpace = "normal";
      span.insertBefore(spacer, child);
    }
  }
} catch (error) {
  // Sometimes KaTeX throws with undefined environments; this is a bug in KaTeX.
  element.innerHTML =
    '<span style="color: rgb(204, 0, 0);">Unknown math error.</span>';
}
```

It's not a beautiful workaround, but hopefully this helps someone with a similar problem.

**Comment by edemaine** [comment]
If this is documented behavior (`white-space: normal` within `white-space: nowrap` can actually break lines?), it seems like something KaTeX could adopt directly. Would be neat to solve this long-standing issue.

**Comment by ekzhang** [comment]
Sounds good. I would be happy to try to contribute a PR for this. Whether this is "documented behavior" - I agree it's somewhat surprising, honestly just tried different things until Chrome was happy with it. The [CSS Text Module Level 3 spec](https://www.w3.org/TR/css-text-3/#white-space-property) has a section on the `white-space` property:

normal
This value directs user agents to collapse sequences of white space into a single character (or in some cases, no character). Lines may wrap at allowed soft wrap opportunities, as determined by the line-breaking rules in effect, in order to minimize inline-axis overflow.

**Comment by ekzhang** [comment]
I'm running the dev server at `localhost:7936`, but it only shows display math rendering, and this issue/change specifically is for inline math line breaks. Is there a way I can inspect the inline math rendering?

**Comment by edemaine** [comment]
@ekzhang Yes, modify the URL to `http://localhost:7936/?display=0`.  See `static/main.js` for other available options.

**Comment by ekzhang** [comment]
Apologies for the delay, but I found some time to work on this. Just filed #3133, which should fix the issue - let me know how it looks!

**Comment by fweth** [comment]
Does the philosophy of the KaTeX auto-renderer focus solely on reading and replacing content within the delimiters? I am attempting to narrow down the range of potential solutions. Would it be feasible for a solution to involve the parser scanning the symbols following the delimiter and integrating them into the replaced HTML element?

**Comment by fweth** [comment]
One idea that seems to work is to have the position of the last inline-block element set to absolute and create an empty span element in front of it with the same width (realized as margin). Here is a hand-coded example which works for me on Firefox: https://jsfiddle.net/xjkmuws8/17/

Would it be interesting to work this approach into a script?

**Comment by rupumped** [comment]
Also not claiming that this is a good solution, but in case you want to accomplish this with JS instead of changing the CSS of all the last-child bases, pulling the immediate next character into the base (assuming it's not whitespace) works, too:
```
window.addEventListener('load', () => {
  document.querySelectorAll('span:has(span.katex)').forEach(katexSpan => {
    if (katexSpan.nextSibling && katexSpan.nextSibling.nodeType === Node.TEXT_NODE && !/\s/.test(katexSpan.nextSibling.textContent.charAt(0))) {
      katexSpan.lastChild.lastChild.lastChild.innerHTML+= katexSpan.nextSibling.textContent.charAt(0);
      katexSpan.nextSibling.textContent = katexSpan.nextSibling.textContent.slice(1);
    }
  });
})
```
